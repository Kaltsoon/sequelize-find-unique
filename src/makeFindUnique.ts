import { FindOptions, ModelStatic } from 'sequelize';
import DataLoaderManager from './dataLoaderManager';
import { findRowByWhere, getWhereQuery } from './utils';

// eslint-disable-next-line @typescript-eslint/no-empty-function
const noop = () => {};

interface MakeFindUniqueOptions {
  onLoadBatch: (keys: ReadonlyArray<FindOptions<any>>) => void;
}

const makeFindUnique = <ModelClass extends ModelStatic<any>>(
  Model: ModelClass,
  options: MakeFindUniqueOptions,
) => {
  const onLoadBatch = options?.onLoadBatch ?? noop;

  const batchLoadFn = async (keys: ReadonlyArray<FindOptions<any>>) => {
    onLoadBatch(keys);

    const whereQuery = getWhereQuery(keys);

    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const { where, ...restQueryOptions } = keys[0];

    const rows = await Model.findAll({
      where: whereQuery,
      ...restQueryOptions,
    });

    return keys.map((key) =>
      key.where ? findRowByWhere(rows, key.where) ?? null : null,
    );
  };

  const manager = new DataLoaderManager(batchLoadFn);

  const findUnique = (
    queryOptions: FindOptions<any>,
  ): Promise<InstanceType<ModelClass> | null> => {
    const loader = manager.getLoader(queryOptions);

    return loader.load(queryOptions);
  };

  return findUnique;
};

export default makeFindUnique;
