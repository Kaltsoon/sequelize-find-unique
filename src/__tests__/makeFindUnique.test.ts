import { Sequelize, DataTypes } from 'sequelize';

import makeFindUnique from '../makeFindUnique';

const sequelize = new Sequelize('sqlite::memory:');

const User = sequelize.define('user', {
  id: { type: DataTypes.INTEGER, primaryKey: true, autoIncrement: true },
  username: { type: DataTypes.TEXT, unique: true },
});

const Comment = sequelize.define('comment', {
  id: { type: DataTypes.INTEGER, primaryKey: true, autoIncrement: true },
  userId: { type: DataTypes.INTEGER },
  content: { type: DataTypes.TEXT },
});

const Person = sequelize.define(
  'person',
  {
    id: { type: DataTypes.INTEGER, primaryKey: true, autoIncrement: true },
    firstName: { type: DataTypes.INTEGER },
    lastName: { type: DataTypes.INTEGER },
  },
  {
    indexes: [
      {
        unique: true,
        fields: ['firstName', 'lastName'],
      },
    ],
  },
);

User.hasMany(Comment);
Comment.belongsTo(User);

describe('makeFindUnique', () => {
  beforeEach(async () => {
    await sequelize.sync({ force: true });
  });

  it('batches queries correctly', async () => {
    const onLoadBatch = jest.fn();

    const findUniqueUser = makeFindUnique(User, { onLoadBatch });

    User.bulkCreate([
      { username: 'kalle' },
      { username: 'elina' },
      { username: 'lasse' },
    ]);

    const result = await Promise.all([
      findUniqueUser({
        where: {
          username: 'kalle',
        },
        attributes: ['username', 'id'],
      }),
      findUniqueUser({
        where: {
          username: 'elina',
        },
        attributes: ['username', 'id'],
      }),
      findUniqueUser({
        where: {
          username: 'lasse',
        },
        attributes: ['username', 'id'],
      }),
    ]);

    const { calls } = onLoadBatch.mock;

    expect(onLoadBatch).toHaveBeenCalledTimes(1);
    expect(calls).toMatchSnapshot();
    expect(result).toMatchSnapshot();
  });

  it('batches queries correctly with composite unique columns', async () => {
    const onLoadBatch = jest.fn();

    const findUniquePerson = makeFindUnique(Person, { onLoadBatch });

    Person.bulkCreate([
      { firstName: 'Kalle', lastName: 'Ilves' },
      { firstName: 'John', lastName: 'Doe' },
    ]);

    const result = await Promise.all([
      findUniquePerson({
        where: {
          firstName: 'Kalle',
          lastName: 'Ilves',
        },
        attributes: ['id', 'firstName', 'lastName'],
      }),
      findUniquePerson({
        where: {
          firstName: 'John',
          lastName: 'Doe',
        },
        attributes: ['id', 'firstName', 'lastName'],
      }),
    ]);

    const { calls } = onLoadBatch.mock;

    expect(onLoadBatch).toHaveBeenCalledTimes(1);
    expect(calls).toMatchSnapshot();
    expect(result).toMatchSnapshot();
  });

  it('queries with same include and attributes parameters are in the same batch', async () => {
    const onLoadBatch = jest.fn();

    const findUniqueUser = makeFindUnique(User, { onLoadBatch });

    User.bulkCreate([
      { username: 'kalle', id: 1 },
      { username: 'elina', id: 2 },
      { username: 'lasse', id: 3 },
    ]);

    Comment.bulkCreate([
      { content: 'text 1', userId: 1 },
      { content: 'text 2', userId: 2 },
    ]);

    const result = await Promise.all([
      findUniqueUser({
        where: {
          username: 'kalle',
        },
        attributes: ['username', 'id'],
        include: { model: Comment, attributes: ['id', 'content'] },
      }),
      findUniqueUser({
        where: {
          username: 'elina',
        },
        attributes: ['username', 'id'],
        include: { model: Comment, attributes: ['id', 'content'] },
      }),
      findUniqueUser({
        where: {
          username: 'lasse',
        },
        attributes: ['username', 'id'],
        include: { model: Comment, attributes: ['id', 'content'] },
      }),
    ]);

    const { calls } = onLoadBatch.mock;

    expect(onLoadBatch).toHaveBeenCalledTimes(1);
    expect(calls).toMatchSnapshot();
    expect(result).toMatchSnapshot();
  });

  it('queries with different include or attributes parameters are not in the same batch', async () => {
    const onLoadBatch = jest.fn();

    const findUniqueUser = makeFindUnique(User, { onLoadBatch });

    User.bulkCreate([
      { username: 'kalle', id: 1 },
      { username: 'elina', id: 2 },
    ]);

    Comment.bulkCreate([{ content: 'text 1', userId: 1 }]);

    const result = await Promise.all([
      findUniqueUser({
        where: {
          username: 'kalle',
        },
        attributes: ['username', 'id'],
        include: { model: Comment, attributes: ['id', 'content'] },
      }),
      findUniqueUser({
        where: {
          username: 'elina',
        },
        attributes: ['username'],
      }),
    ]);

    const { calls } = onLoadBatch.mock;

    expect(onLoadBatch).toHaveBeenCalledTimes(2);
    expect(calls).toMatchSnapshot();
    expect(result).toMatchSnapshot();
  });

  it('queries with different where parameters are not in the same batch', async () => {
    const onLoadBatch = jest.fn();

    const findUniqueUser = makeFindUnique(User, { onLoadBatch });

    User.bulkCreate([
      { username: 'kalle', id: 1 },
      { username: 'elina', id: 2 },
    ]);

    Comment.bulkCreate([{ content: 'text 1', userId: 1 }]);

    const result = await Promise.all([
      findUniqueUser({
        where: {
          username: 'kalle',
        },
        attributes: ['username', 'id'],
      }),
      findUniqueUser({
        where: {
          id: 2,
        },
        attributes: ['username', 'id'],
      }),
    ]);

    const { calls } = onLoadBatch.mock;

    expect(onLoadBatch).toHaveBeenCalledTimes(2);
    expect(calls).toMatchSnapshot();
    expect(result).toMatchSnapshot();
  });
});
