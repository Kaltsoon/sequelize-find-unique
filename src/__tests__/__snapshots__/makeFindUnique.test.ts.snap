// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`makeFindUnique batches queries correctly 1`] = `
Array [
  Array [
    Array [
      Object {
        "attributes": Array [
          "username",
          "id",
        ],
        "where": Object {
          "username": "kalle",
        },
      },
      Object {
        "attributes": Array [
          "username",
          "id",
        ],
        "where": Object {
          "username": "elina",
        },
      },
      Object {
        "attributes": Array [
          "username",
          "id",
        ],
        "where": Object {
          "username": "lasse",
        },
      },
    ],
  ],
]
`;

exports[`makeFindUnique batches queries correctly 2`] = `
Array [
  Object {
    "id": 1,
    "username": "kalle",
  },
  Object {
    "id": 2,
    "username": "elina",
  },
  Object {
    "id": 3,
    "username": "lasse",
  },
]
`;

exports[`makeFindUnique batches queries correctly with composite unique columns 1`] = `
Array [
  Array [
    Array [
      Object {
        "attributes": Array [
          "id",
          "firstName",
          "lastName",
        ],
        "where": Object {
          "firstName": "Kalle",
          "lastName": "Ilves",
        },
      },
      Object {
        "attributes": Array [
          "id",
          "firstName",
          "lastName",
        ],
        "where": Object {
          "firstName": "John",
          "lastName": "Doe",
        },
      },
    ],
  ],
]
`;

exports[`makeFindUnique batches queries correctly with composite unique columns 2`] = `
Array [
  Object {
    "firstName": "Kalle",
    "id": 1,
    "lastName": "Ilves",
  },
  Object {
    "firstName": "John",
    "id": 2,
    "lastName": "Doe",
  },
]
`;

exports[`makeFindUnique queries with different include or attributes parameters are not in the same batch 1`] = `
Array [
  Array [
    Array [
      Object {
        "attributes": Array [
          "username",
          "id",
        ],
        "include": Object {
          "attributes": Array [
            "id",
            "content",
          ],
          "model": [Function],
        },
        "where": Object {
          "username": "kalle",
        },
      },
    ],
  ],
  Array [
    Array [
      Object {
        "attributes": Array [
          "username",
        ],
        "where": Object {
          "username": "elina",
        },
      },
    ],
  ],
]
`;

exports[`makeFindUnique queries with different include or attributes parameters are not in the same batch 2`] = `
Array [
  Object {
    "comments": Array [
      Object {
        "content": "text 1",
        "id": 1,
      },
    ],
    "id": 1,
    "username": "kalle",
  },
  Object {
    "username": "elina",
  },
]
`;

exports[`makeFindUnique queries with different where parameters are not in the same batch 1`] = `
Array [
  Array [
    Array [
      Object {
        "attributes": Array [
          "username",
          "id",
        ],
        "where": Object {
          "username": "kalle",
        },
      },
    ],
  ],
  Array [
    Array [
      Object {
        "attributes": Array [
          "username",
          "id",
        ],
        "where": Object {
          "id": 2,
        },
      },
    ],
  ],
]
`;

exports[`makeFindUnique queries with different where parameters are not in the same batch 2`] = `
Array [
  Object {
    "id": 1,
    "username": "kalle",
  },
  Object {
    "id": 2,
    "username": "elina",
  },
]
`;

exports[`makeFindUnique queries with same include and attributes parameters are in the same batch 1`] = `
Array [
  Array [
    Array [
      Object {
        "attributes": Array [
          "username",
          "id",
        ],
        "include": Object {
          "attributes": Array [
            "id",
            "content",
          ],
          "model": [Function],
        },
        "where": Object {
          "username": "kalle",
        },
      },
      Object {
        "attributes": Array [
          "username",
          "id",
        ],
        "include": Object {
          "attributes": Array [
            "id",
            "content",
          ],
          "model": [Function],
        },
        "where": Object {
          "username": "elina",
        },
      },
      Object {
        "attributes": Array [
          "username",
          "id",
        ],
        "include": Object {
          "attributes": Array [
            "id",
            "content",
          ],
          "model": [Function],
        },
        "where": Object {
          "username": "lasse",
        },
      },
    ],
  ],
]
`;

exports[`makeFindUnique queries with same include and attributes parameters are in the same batch 2`] = `
Array [
  Object {
    "comments": Array [
      Object {
        "content": "text 1",
        "id": 1,
      },
    ],
    "id": 1,
    "username": "kalle",
  },
  Object {
    "comments": Array [
      Object {
        "content": "text 2",
        "id": 2,
      },
    ],
    "id": 2,
    "username": "elina",
  },
  Object {
    "comments": Array [],
    "id": 3,
    "username": "lasse",
  },
]
`;
